rhel/vex/parser.go

```diff
@@ -183,31 +183,46 @@ type creator struct {
 	rc                 *repoCache
 }
 
-// WalkRelationships attempts to resolve a relationship until we have a package product_id and
-// a repo product_id. Relationships can be nested. If the pkgID and the repoID are the same we
-// either found no relationship or a relationship where both ends are pointing to the same
-// product_id, either way we don't have enough data to create a vulnerability.
-func walkRelationships(productID string, doc *csaf.CSAF) (string, string, error) {
-	pkgID, repoID := extractProductNames(productID, productID, doc)
-	if pkgID == repoID {
-		return "", "", fmt.Errorf("could not extract a distict pkgID and repoID from %q", productID)
+// WalkRelationships attempts to resolve a relationship until we have a package product_id,
+// a repo product_id and possibly a package module product_id. Relationships can be nested.
+// If we don't get an initial relationship or we don't get two component parts we cannot
+// create a vulnerability. We never see more than 3 components in the wild but if we did
+// we'd assume the component next to the repo product_id is the package module.
+func walkRelationships(productID string, doc *csaf.CSAF) (string, string, string, error) {
+	comps := []string{}
+	prodRel := doc.FindRelationship(productID, "default_component_of")
+	if prodRel == nil {
+		return "", "", "", fmt.Errorf("cannot determine initial relationship for %q", productID)
+	}
+	comps = extractProductNames(prodRel.ProductRef, prodRel.RelatesToProductRef, comps, doc)
+	switch {
+	case len(comps) == 2:
+		// We have a package and repo
+		return comps[0], "", comps[1], nil
+	case len(comps) > 2:
+		// We have a package, module and repo
+		return comps[0], comps[len(comps)-2], comps[len(comps)-1], nil
+	default:
+		return "", "", "", fmt.Errorf("cannot determine relationships for %q", productID)
 	}
-	return pkgID, repoID, nil
 }
 
-// ExtractProductNames recursively looks up the package product_id and the repo product_id.
-// The assumtion is that the repo is always the last found relates_to_product_reference and the
-// package is the last found product_reference.
-func extractProductNames(prodRelID string, repoRelID string, c *csaf.CSAF) (string, string) {
-	prodRel := c.FindRelationship(prodRelID, "default_component_of")
+// ExtractProductNames recursively looks up product_id relationships and adds them to a
+// component slice in order.
+func extractProductNames(right, left string, comps []string, c *csaf.CSAF) []string {
```

Comment:
It's a little confusing that `right` is the leftmost argument and `left` is to the right of `right`. Can we change this?

Reply:
It's referring to the sides of productID string, i.e 
```
repo:mod:pkg
```
 which will then become:
```
repo:mod    pkg
```
and then:
```
repo    mod     pkg
```
The order of the comps is reversed when they're returned so I can see that might be confusing, what would you suggest to make it less so? `prodRef` and `relatesToProdRef`?

Reply:
yeah maybe something descriptive like `prodRef ` and `relatesToProdRef ` like you suggested would be good.